liota/dccs/aws_iot.py

```diff
@@ -0,0 +1,148 @@
+        """
+        This method creates Parent-Child relationship.  Supported relationships are:
+
+               EdgeSystem
+                   |                                      EdgeSystem
+                Device                   (or)                |
+                   |                                    RegisteredMetric
+             RegisteredMetric
+
+        However, A single EdgeSystem can have multiple child Devices and a each Device can have
+        multiple child Metrics.
+
+        :param reg_entity_parent: EdgeSystem or Device Object
+        :param reg_entity_child:  Device or RegisteredMetric Object
+        :return: None
+        """
+        if not isinstance(reg_entity_parent, EdgeSystem) and not isinstance(reg_entity_parent, Device):
+            raise TypeError("reg_entity_parent should either be an EdgeSystem or a Device")
+        if isinstance(reg_entity_parent, Device):
+            if not isinstance(reg_entity_parent.parent, EdgeSystem):
+                raise TypeError("Parent Child relationship between EdgeSystem and Device must be created.")
+        if not isinstance(reg_entity_child, Device) and not isinstance(reg_entity_child, RegisteredMetric):
+            raise TypeError("reg_entity_child should either be a Device or a RegisteredMetric")
+        reg_entity_child.parent = reg_entity_parent
+
+    def _get_publish_topic(self, reg_metric):
+        """
+        :param reg_metric: RegisteredMetric Object
+        :return: MQTT publish topic based on Parent Child relationship
+        """
+        if not isinstance(reg_metric, RegisteredMetric):
+            raise TypeError("RegisteredMetric is expected")
+
+        def extract_topic(entity):
+            """
+            Recursive Function for extracting MQTT topic
+            :param entity: EdgeSystem or Device or RegisteredMetric Object
+            :return: MQTT Topic
+            """
+            if isinstance(entity, EdgeSystem):
+                return entity.name + '/'
+            elif isinstance(entity, Device):
+                return extract_topic(entity.parent) + entity.name + '/'
+            elif isinstance(entity, RegisteredMetric):
+                return extract_topic(entity.parent) + entity.ref_entity.name
+            else:
+                raise TypeError("Problem occurred while extracting publish topic")
+
+        return extract_topic(reg_metric)
+
+    def _format_data(self, reg_metric):
+        """
+        :param reg_metric: Registered Metric Object
+        :return: Payload in JSON format
+        """
+        list = []
+        met_cnt = reg_metric.values.qsize()
+        if met_cnt > 0:
+            for _ in range(met_cnt):
+                m = reg_metric.values.get(block=True)
+                if m is not None:
+                    list.append(OrderedDict([('value', m[1]), ('timestamp', m[0])]))
+        payload = OrderedDict()
+        payload['metric_name'] = reg_metric.ref_entity.name
```

Comment:
Its good to add "Device Name" and "Gateway Name" as part of payload.  This will help to carry forward important fields along with the message. (Helps in doing aggregation, analytics)
Today this information is available from TopicName but this wont be carried forward with the message.